### PR TITLE
Fetch attribute values independently in MetadataViewer

### PR DESCRIPTION
**Pull request by axelboc**
I'm trying to do the refactoring for #807 in a safe way, so I've added the new behaviour on top of the current one. Therefore, for now, the attribute values:

- are fetched twice: first as part of `getEntity` and now also in `getAttrValues`, which is invoked when inspecting a dataset;
- are stored in two places: in the `Entity` objects, and in the new `attrValuesStore`.

I've also limited the refactoring to the `MetadataViewer` for now. The next step will be do deal with NeXus, and the one after that to remove the old attribute-fetching behaviour.

attr-loading

You can see the new suspense loader in the `MetadataViewer` above with HSDS but not with H5Grove, because in `H5GroveApi`, `fetchAttrValues` keeps its own cache (and `getAttrValues` is always called after `getEntity`, so `getAttrValues` hits the cache). Hopefully at the end of this refactoring, this caching will no longer be required.

### CODE REVIEW

File: packages/app/src/metadata-viewer/AttributesInfo.tsx [hunk 1]
```diff
@@ -13,28 +14,21 @@ const FOLLOWABLE_ATTRS = new Set([
...
   return (
-    <table className={styles.table}>
-      <thead>
-        <tr>
-          <th scope="col" colSpan={2}>
-            Attributes
-          </th>
-        </tr>
-      </thead>
-      <tbody>
```
Comment (axelboc) [inline packages/app/src/metadata-viewer/AttributesInfo.tsx:36]: I've extracted the `table` wrapper in a separate component for re-use.

File: packages/app/src/metadata-viewer/AttributesInfo.tsx [hunk 2]
```diff
@@ -13,28 +14,21 @@ const FOLLOWABLE_ATTRS = new Set([
...
-function AttributesTable(props: Props) {
-  const { attributes, onFollowPath } = props;
+function AttributesInfo(props: Props) {
+  const { entity, onFollowPath } = props;
 
-  if (attributes.length === 0) {
-    return null;
-  }
+  const { attrValuesStore } = useContext(ProviderContext);
+  const attrValues = attrValuesStore.get(entity);
```
Comment (axelboc) [inline packages/app/src/metadata-viewer/AttributesInfo.tsx:25]: This is from where the new behaviour is triggered.

File: packages/app/src/providers/hsds/hsds-api.ts [hunk 3]
```diff
@@ -161,12 +185,9 @@ export class HsdsApi extends ProviderApi {
...
 
-    const attrsPromises = data.attributes.map(async (attr) => {
-      const { data } = await this.client.get<HsdsAttributeResponse>(
-        `/${entityCollection}/${entityId}/attributes/${attr.name}`
-      );
-      return data;
-    });
+    const attrsPromises = data.attributes.map(async (attr) =>
+      this.fetchAttributeWithValue(entityCollection, entityId, attr.name)
+    );
```
Comment (axelboc) [inline packages/app/src/providers/hsds/hsds-api.ts:190]: The `fetchAttributes` method currently fetches all the attributes' metadata at once, and then each attribute's value one at a time like this. At the end of this refactoring, this look will be gone from here and only occur in `getAttrValues`.

File: packages/app/src/providers/hsds/hsds-api.ts [hunk 4]
```diff
@@ -184,8 +205,20 @@ export class HsdsApi extends ProviderApi {
     return data.value;
   }
 
+  private async fetchAttributeWithValue(
+    entityCollection: HsdsCollection,
```
Comment (axelboc) [inline packages/app/src/providers/hsds/hsds-api.ts:209]: The endpoint needs to know the HSDS entity's `collection`, which is like the `type` but with an `s`... So far we were discarding this information which comes from the link resolution phase. Instead of trying to guess the collection from the type, which is tricky for unresolved entities... I decided to keep the `collection` around like we do for the `id`. That's why you see it a bit everywhere in the diff.

File: packages/shared/src/mock/utils.ts [hunk 5]
```diff
@@ -1,28 +1,25 @@
...
-function isMockDataset(entity: Entity): entity is MockDataset {
-  return isDataset(entity) && 'value' in entity;
-}
-
-export function assertMockDataset(
-  entity: Entity
-): asserts entity is MockDataset {
-  if (!isMockDataset(entity)) {
+export function assertMockDataset<S extends Shape, T extends DType>(
+  dataset: Dataset<S, T>
```
Comment (axelboc) [inline packages/shared/src/mock/utils.ts:20]: Unrelated change: the guard now accepts a dataset instead of an entity, which means that callers just need to make sure to call `assertDataset` beforehand if they need to (which is not always the case).

File: packages/shared/src/models-hdf5.ts [hunk 6]
```diff
@@ -32,11 +32,6 @@ export interface Dataset<S extends Shape = Shape, T extends DType = DType>
   rawType?: unknown;
 }
 
-export interface MockDataset<S extends Shape = Shape, T extends DType = DType>
-  extends Dataset<S, T> {
-  value: unknown;
-}
-
```
Comment (axelboc) [inline packages/shared/src/models-hdf5.ts:39]: dead code

File: packages/app/src/metadata-viewer/MetadataViewer.tsx [hunk 7]
```diff
@@ -22,19 +31,42 @@ function MetadataViewer(props: Props) {
...
+        <MetadataTable title="Attributes">
+          <ErrorBoundary
+            resetKeys={[path]}
+            FallbackComponent={AttributesErrorFallback}
+          >
+            <Suspense
+              fallback={
+                <tr>
+                  <td className={styles.fallback}>Loading...</td>
+                </tr>
```
Comment (loichuder) [inline packages/app/src/metadata-viewer/MetadataViewer.tsx:53]: Might be worth putting in its own component ? :man_shrugging:

**Review by loichuder — APPROVED** [review]
Well-thought-out :+1:
